library/x509_crt.c

```diff
@@ -2951,23 +2951,147 @@ static int x509_crt_verify_chain(
     }
 }
 
+/*
+ * parse ipv4 address from canonical string form into bytes.
+ * return 1 if success, 0 otherwise (consistent with inet_pton)
+ */
+static int parse_ipv4( const char *h, char *addr )
+{
+    unsigned char ip[4];
+    int i;
+    const char *strt = h;
+    char *endp;
+    unsigned long v;
+    for( i = 0; i < 4; i++ )
+    {
+        v = strtoul( strt, &endp, 10 );
+        if( endp == strt || v > 255 )
+        {
+            return ( 0 );
+        }
+
+        if( i < 3 && *endp != '.' )
+        {
+            return ( 0 );
+        }
+        ip[i] = v & (unsigned char)0xFF;
+        strt = endp + 1;
+    }
+
+    memcpy( addr, ip, sizeof( ip ) );
+    return ( 1 );
+}
+
+/*
+ * parse ipv6 address from canonical string form into bytes.
+ * return 1 if success, 0 otherwise (consistent with inet_pton)
+ */
+static int parse_ipv6( const char *h, size_t hlen, char *addr )
```

Comment:
Minor: I tested this function with "aaaa::aabg" and received success, with returned address as "aaaa::0aab".
I would expect this function to return false for such input, but since this is a static function, which the input is mostly controlled, and output compared to ip, and doesn't seem to me like a security issue, I wouldn't block.
However, It is worth to add tests with other invalid addresses

Reply:
added additional checks for invalid ipv6 in latest commit